### ISSUE TITLE
bug: issue with `--extended` or `-@`

### ISSUE THREAD

**Issue by origamirabbit** [eza-community/eza#425]
Just found this community fork support and really appreciate your work!

I tried to swap the alias from `exa` to `eza` in `.zshrc`, but one of the alias seems not working in eza but working in exa. (see **The command-line arguments you are using**)

- The version of eza being used (`eza --version`)

```
eza --version
eza - A modern, maintained replacement for ls
v0.13.1 [+git]
https://github.com/eza-community/eza
```

- The command-line arguments you are using

```
eza -lbhHigUmuSaZ@ --time-style=long-iso --git --color-scale
```
I apologise if the arguments I used are incorrectly configured and please correct me if I am wrong, cheers!

- Your operating system and hardware platform

```
macOS Ventura v13.6
MacBook Pro 2019 (Intel)
```

---

**Comment by ariasuni** [comment]
It works on my machine, do you have this crash in every directory you try? What if you remove some options, does it still crash?

**Comment by ariasuni** [comment]
I read a bit too fast the bug description, I’m on Linux so that may be a macOS-only issue. Some options aren’t «triggered» when using `--long`, in this case it means that if you try to use `--extended` with `--long` it won’t even try to read the extended attributes (in `eza --help`, `--extended` is listed in the `LONG VIEW OPTIONS` to indicate you that).

**Comment by elvilio** [comment]
can confirm segmentation fault every time in `~/`
executing `eza -la@` -> `[1]    20342 segmentation fault  eza -la@`
`eza -l@` -> correct behaviour
`eza -a@` -> correct behaviour
`eza -la` -> correct bahviour

happens in `~/`, `/tmp/`, `/Applications`, `~/Applications` (which its strange since I only have 3 automator applications)
doesn't happen in `/dev`, `opt`, and a bunch of other directories i have tried

an error message would be much better than segmentation fault even if some options are incompatible

using iTerm but happens in the default terminal too
using zsh but happens from bash too
`rustc 1.72.1 (d5c2e9c34 2023-09-13)`
macOS Sonoma 14.0 (23A344)

**Comment by jcoetzee** [comment]
Also bumped into this today, also on certain folders (e.g. `~/`)

```
❯ eza -l@
fish: Job 1, 'eza -l@' terminated by signal SIGSEGV (Address boundary error)
```

```
❯ eza --version
eza - A modern, maintained replacement for ls
v0.14.1 [+git]
https://github.com/eza-community/eza
```

**Comment by jcoetzee** [comment]
Sorry macOS as well. `Sonoma 14.0 (23A344)`

**Comment by cobaltdr** [comment]
Hello. First of all, thank you for the fork. And sorry, I'm a bit late to the party.

I can confirm this bug on iTerm2 3.4.21, Macos Sonoma 13.5.1, zsh 5.9 (arm-apple-darwin22.1.0) from homebrew.

In my case, however, this does not happen in every directory, which makes me think it has something to do with the sequence of unprintable characters being the contents of the extended attributes. In fact, I was logging onto Github to request a feature to limit the characters listed by --extended/-@ and found this bug.

There is also another possibly related bug with --icons where I was using a font without the required glyphs and it would crash in particular directories.

**Comment by oscillik** [comment]
macOS 14.1 (23B74), zsh 5.9

getting Segmentation Fault when running `eza -l@`. This occurs in iTerm2 as well as Terminal.

**Comment by cafkafk** [comment]
Wonder if this is at all related to #576

**Comment by oscillik** [comment]
I've not idea, but I can confirm that the latest build of`exa` does not exhibit the same issue, so something has happened somewhere!

**Comment by unixwizard107** [comment]
Sorry for the duplicate (#603).   I did not see the earlier issue. Please let me know if I can do something like sending any other info than I already put in my report to help you find the error  - please let me know.   While I ran into on my Arm-based  Mac, I can try it on an old Intel based  system also if that helps you any.

**Comment by Tyrubias** [comment]
I've been looking into this issue. Like others have commented, I believe this issue is because `eza` is trying to print extended attribute information that is not valid Unicode. Specifically, when painting the `TextCell` for a specific extended attribute, `DisplayWidth::from(input: &'a str)` segfaults when using the `unicode-width` crate.  

On `eza v0.15.2` on ARM macOS v14.1.1, the error can be reproduced with the following steps.

1. Find a directory where this issue doesn't occur (e.g. `~/Downloads`).
2. Create a new, empty directory (e.g. `my_test_dir`).
3. Create a file in this directory (e.g. `touch my_test_file`).
4. Assuming the file is named `$FILE`, run the following command
```
xattr -wx com.apple.macl 00035D550CFA9FB5409F8593FF64E35FC29F0004F45A7ABDB8E14740AEBB584FDA0BF57100035D550CFA9FB5409F8593FF64E35FC29F0004F45A7ABDB8E14740AEBB584FDA0BF571 $FILE
```

The `xattr -wx` command is the important part; it adds the `com.apple.macl` extended attribute with the given value to whatever file or directory is specified. This extended attribute is used as part of Apple's TCC mechanism, specifically the [per-file](https://blog.xpnsec.com/we-need-to-talk-about-macl/) [user intent](https://eclecticlight.co/2020/01/30/quarantine-sip-and-macl-macos-per-file-security-controls/) process. Theoretically, this should just be an UUID, but something about it makes it invalid Unicode.

**Comment by Tyrubias** [comment]
After judicious use of `git bisect`, I believe the offending commit is afeac47c13f81213b48c1d3b317ef7862061e14d

**Comment by daviessm** [comment]
There's a lot of `unsafe` code in there, I think we at least need some [`// SAFETY:`](https://std-dev-guide.rust-lang.org/policy/safety-comments.html) comments explaining how we guarantee that the block is safe, on top of actually making it safe and not just assuming it's valid UTF-8.

**Comment by Tyrubias** [comment]
I agree. As a bandage fix we could switch to using [String::from_utf8_lossy](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy) at the risk of printing the ugly Unicode replacement character.

**Comment by cfxegbert** [comment]
Can you please check in v0.16.1 or later

**Comment by niklasbogensperger** [comment]
Not OP of the issue but eza v0.16.2 installed via Homebrew now works again with this flag on my MacBook (MacBookPro14,3; macOS Ventura 13.6.1) and displays everything the way it should.
(checked with `eza --long --extended`, `eza --long --all --extended`, `eza --long --all --all --extended`, and `eza --all --all --long --group-directories-first --classify --extended --header --time-style=long-iso` in different directories)

**Comment by origamirabbit** [comment]
Can confirm that running `eza -lbhHigUmuSa@ --time-style=long-iso --git --color-scale` will no longer get segmentation fault error with eza v0.16.2.
I noticed the only difference is the last part of the Name column between eza and original exa.

- exa will display the length of the value for the extended item in the list.
`├── com.apple.FinderInfo (len 32)`

- eza will display the value itself instead. (which is better I guess)
`├── com.apple.FinderInfo: "        @\u{10}"`

I think we could close this issue for now.

**Comment by cfxegbert** [comment]
> * exa will display the length of the value for the extended item in the list.
>   `├── com.apple.FinderInfo (len 32)`
> * eza will display the value itself instead. (which is better I guess)
>   `├── com.apple.FinderInfo: "        @\u{10}"`

I try to detect "printable" attributes now and display them.  If it is a binary attribute and under 16 bytes I will display a hex dump else I display the length.  I also decode `com.apple.lastuseddate`, `com.apple.macl`, and binary plist values.
